mpb/fields.c

```diff
@@ -569,13 +569,27 @@ static real get_val(int ix, int iy, int iz,
 #endif
 
 #ifdef HAVE_MPI
-     CHECK(0, "get-*-point not yet implemented for MPI!");
-#else
-     if (conjugate)
-	  return -data[(((ix * ny) + iy) * nz + iz) * stride];
-     else
-	  return data[(((ix * ny) + iy) * nz + iz) * stride];
+	 /* due to real-space xy=>yx transposition in MPI configuration, we need to
+	    do a little extra dancing here; see details e.g. in XYZ_LOOP macro */
+     int local_ny = mdata->local_ny; /* dim of local process over y-indices */
+     int local_y_start = mdata->local_y_start;
+     int local_iy = iy - local_y_start;
+	 real val = 0; /* reduce local processes over this variable later */
+
+	 /* check if local_iy is in the current process' data block */
+     if (local_iy >= 0 && local_iy < local_ny) { 
+         val = data[(((local_iy * nx) + ix) * nz + iz) * stride]; /* note transposition in x and y indices */
+	 }
+	 mpi_allreduce_1(&val, real, SCALAR_MPI_TYPE, MPI_SUM, mpb_comm); 
```

Comment:
We could actually call this in `interp_val` on the result of the bilinear interpolation (this is the only function that calls `get_val`), rather than in `get_val`, to save a factor of 8 in the number of `allreduce` calls.

Reply:
Ah, nice. Just to check my understanding, you mean we could instead reduce over the operations here,
https://github.com/NanoComp/mpb/blob/0d886cf7e638ce690a9b24060c265fcdc99d305b/mpb/fields.c#L628-L631
and then return the reduced value from `interp_val`? I.e. `get_val` would then be returning a process-specific value when run with MPI?

Reply:
Yes. Do the reduce on the interpolated value.

This gives the same result because the interpolation is a linear operation.

Reply:
Okay, thanks! (my uncertainty was just whether MPI allows a function to return a process-specific value or if it has to be reduced before returning; makes sense that it doesn't need that)

Reply:
Functions can do whatever they want.  MPI is just a library providing subroutines for a bunch of running processes to communicate with one another — each process is an independent program that can do independent computations.